### DISCUSSION TITLE
Dependency Overrides in startup event gives TypeError: 'Depends' object is not callable

### DISCUSSION THREAD

**Discussion by timbmg** [fastapi/fastapi#9134] [Questions]
How can dependencies in the startup event be overwritten for testing?

Below a minimal example. When executing the test, `Model` is a `Depends` object and not a `MagicMock`. When putting the depends in normal endpoints it works just fine.

api.py
```python 
from my_package.model import Model
from fastapi import Depends, FastAPI
app = FastAPI()

def model_provider():
    return Model

model = None

@app.on_event("startup")
def startup(Model = Depends(model_provider)):
    global model
    model = Model()

@app.get("/prediction")
def predict(x):
    return model.predict(x)

```

tests/test_api.py
````python
from my_package import api
from unittest.mock import MagicMock
from fastapi.testclient import TestClient

def test_predict():
    model_mock = MagicMock()
    model_mock.predict = MagicMock(return_value=lambda x: x)
    api.app.dependency_overrides[api.model_provider] = lambda: model_mock
    with TestClient(api.app) as client:
        prediction = client.get("/prediction?x=1")
        assert prediction.text == "1"

```

**Comment by wichert** [comment]
The documents seems to hint that you can only use `Depends` for request functions. Perhaps this could be extended to app events as well?

**Comment by oxfn** [comment]
+1 this would be great to use dependencies in event handlers. Now it is Starlette events which know nothing about FastAPI DI

**Comment by Faylixe** [comment]
+1, the DI is a very important design feature of FastAPI that should be usable in event handler IMO.

Unless already planned for future release, I would like to contribute to this, but this doesn't seems that easy to implement it in a clean way. By reading DI implementation currently used, I realize that it is deeply connected to request / response game, with no separation of concern allowing to just refactor `fastapi.dependencies.utils.solve_dependencies` function to handle an event only use case.

A quick intuition is that it should have a more granular `Dependent` / `Depends` hierarchy like having a base `Dependent` definition without any `Body`, `Form`, `Request`, `Response`, etc ... support and adding those to a child `RequestDependent` before to finally break solver function to support either both use case or smartly trigger errors when unsupported request deps are required in an "non request" context.

But maybe there is a simpler way ?

**Comment by adriangb** [comment]
@Faylixe I took a look, and I agree with your assessment. It looks doable without breaking changes, but yeah a _lot_ of refactoring would be needed.

**Comment by f-fuchs** [comment]
hey, are there any updates on this feature. still does not seem to work for me on version `0.100.1`
